README.md
### Prathamesh Shanbhag's Portfolio Website.🛠



##### I create interactive experiences within the Browser.✨
---
### Main-Page
![Website-Dynamic-Gif](./static/assets/Website-Landing.gif)

### Click-Interaction

![Website-ClickInteraction](./static/assets/Website-ClickInteraction.gif)

### Bloom-Animation

![SocialMedia_IconsPanel_Bloom-Animation](./static/assets/SocialMedia_IconsPanel_Bloom-Animation.gif)


CHANGELOG.md
## [1.4.6](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/compare/v1.4.5...v1.4.6) (2023-03-25)


### Bug Fixes

* Updated Instagram Username & Links ([cdfc97c](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/cdfc97c9b1e8a1f5ebc87893d23178a6e901f08f))



## [1.4.5](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/compare/v1.4.4...v1.4.5) (2022-08-14)


### Bug Fixes

* Updated Community GA to fix broken text & links ([ec52324](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/ec52324b8196359119b972fcdcdc4c86c45fd825))



## [1.4.4](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/compare/v0.5.0...v1.4.4) (2022-08-14)


### Bug Fixes

* Aspect Ratio Issue on Specific Mobile devices ([f89e256](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/f89e2569f0fe09524f949d6b00615cbb1065e919))
* Bad code setting Links_Container Width to 0% ([2a4e9aa](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/2a4e9aaa55b2d1b6b60aca6f880aa5a51cf77b3d))
* Decreased Icon Size to center align icons ([f24dca3](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/f24dca398a4dba062cea3a8027f48a2d157d8de1))
* Fixed Icons Flare Out Animation for Small Devices ([dfa8d66](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/dfa8d6675a0fb070001071773da547b60f7017e6))
* Further Decreased Icon Distance To Fix all Icons on Smaller Devices ([365cb34](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/365cb341349635346df10ad6f923098c7dc7249f))
* Reverting Changes + Increasing Icon size / Decreasing Distance b/wthem ([51d00d3](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/51d00d3bac6f273fc7e57551aad9a3687a122eb1))
* Text Alignment + Font Size Fixed + Distance b/w Icons decreased for Mobile Devices ([5b3ca24](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/5b3ca24b5eee3fd7fed5f268268c3053c67aa26c))
* Text Overlapping Model On Small Devices Fixed ([8734e46](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/commit/8734e46d6cc033e07d16f6317debd10807bc8d17))



## [1.4.3](https://github.com/Prathamesh-Shanbhag/Portfolio-Website/compare/v0.4.6...v1.4.3) (2022-08-09)





src/shaders/Scan.js
export default () => ({
  uniforms: {
    tDiffuse: { value: null },
    tDepth: { value: null },
    scan: { value: 0 },
    cameraNear: { value: 0 },
    cameraFar: { value: 1 },
  },
  vertexShader: `
    varying vec2 vUv;
    void main() {
      vUv = uv;
      gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
    }
  `,
  fragmentShader: `
    #include <packing>

    varying vec2 vUv;

    uniform sampler2D tDiffuse;
    uniform sampler2D tDepth;

    uniform float cameraNear;
    uniform float cameraFar;

    uniform float scan;

    float readDepth(sampler2D depthSampler, vec2 coord) {
      float fragCoordZ = texture2D(depthSampler, coord).x;
      float viewZ = perspectiveDepthToViewZ(fragCoordZ, cameraNear, cameraFar);
      return viewZToOrthographicDepth(viewZ, cameraNear, cameraFar);
    }
    
    void main()
    {
      vec4 color = texture2D(tDiffuse, vUv);
      float depth = readDepth(tDepth, vUv);

      float scanClamp = clamp(scan, 0.0, 1.0);

      color.rgb += pow(1.0 - abs(depth-scanClamp), 140.0) * (1.0 - pow(depth, 10.0));
      

      gl_FragColor = color;
    }
  `,
})


src/shaders/ASCII.js
import * as THREE from 'three'

export default () => ({
  uniforms: {
    tLowRes: { value: null },
    tFont: { value: null },
    tDepth: { value: null },
    fontCharTotalCount: { value: 0 },
    fontCharCount: { value: new THREE.Vector2(1, 1) },
    fontCharSize: { value: new THREE.Vector2(1, 1) },
    renderCharCount: { value: new THREE.Vector2(1, 1) },
    renderCharSize: { value: new THREE.Vector2(1, 1) },
    cameraNear: { value: 0 },
    cameraFar: { value: 1 },
  },
  vertexShader: `
    varying vec2 vUv;
    void main() {
      vUv = uv;
      gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
    }
  `,
  fragmentShader: `
   #include <packing>

    varying vec2 vUv;
    uniform sampler2D tLowRes;
    uniform sampler2D tDepth;
    uniform sampler2D tFont;

    uniform float cameraNear;
    uniform float cameraFar;

    uniform float fontCharTotalCount;
    uniform vec2 fontCharSize;
    uniform vec2 fontCharCount;

    uniform vec2 renderCharCount;
    uniform vec2 renderCharSize;

    float readDepth(sampler2D depthSampler, vec2 coord) {
      float fragCoordZ = texture2D(depthSampler, coord).x;
      float viewZ = perspectiveDepthToViewZ(fragCoordZ, cameraNear, cameraFar);
      return viewZToOrthographicDepth(viewZ, cameraNear, cameraFar);
    }

    void main()
    {
      vec2 roundedUv = vec2(
        floor(vUv.x * renderCharCount.x),
        floor(vUv.y * renderCharCount.y)
      ) * renderCharSize;

      float depth = readDepth(tDepth, roundedUv);
      vec4 color = texture2D(tLowRes, roundedUv);
    
      float charIndex = depth * fontCharTotalCount;

      vec2 fontuv = vec2(
        mod(vUv.x, renderCharSize.x),
        mod(vUv.y, renderCharSize.y)
      ) * renderCharCount * fontCharSize + vec2(
        floor(mod(charIndex, fontCharCount.x)) * fontCharSize.x,
        floor(charIndex * fontCharSize.x) * fontCharSize.y
      );
      
      gl_FragColor = texture2D(tFont, fontuv) * color;
    }
  `,
})


src/shaders/Ripple.js
import { Vector2 } from 'three'

export default () => ({
  uniforms: {
    tDiffuse: { value: null },
    tRipple: { value: null },
    distort: { value: new Vector2(0.001, 0.001) }
  },
  vertexShader: `
    varying vec2 vUv;
    void main() {
      vUv = uv;
      gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
    }
  `,
  fragmentShader: `
    varying vec2 vUv;

    uniform sampler2D tDiffuse;
    uniform sampler2D tRipple;

    uniform vec2 distort;
    
    void main()
    {
      vec3 ripple = texture2D(tRipple, vUv).rgb;

      vec2 distortOffset = normalize(vUv.xy - ripple.xy) * ripple.b + distort;

      gl_FragColor = vec4(
        ripple.b + texture2D(tDiffuse, vUv + distortOffset).r ,
        ripple.b + texture2D(tDiffuse, vUv).g,
        ripple.b + texture2D(tDiffuse, vUv - distortOffset).b,
        1.0
      );
    }
  `,
})


src/shaders/VertexLitParticle.js
import { UniformsUtils, UniformsLib } from 'three'

export default () => ({
  lights: true,
  uniforms: UniformsUtils.merge([
    UniformsLib['lights'],
    {
      pointSize: { value: 2.0 },
      decayModifier: { value: 1.0 },
    },
  ]),
  vertexShader: `
    #include <common>

    struct PointLight {
      vec3 position;
      vec3 color;
      float distance;
      float decay;
      int shadow;
      float shadowBias;
      float shadowRadius;
      vec2 shadowMapSize;
      float shadowCameraNear;
      float shadowCameraFar;
    };

    uniform PointLight pointLights[NUM_POINT_LIGHTS];
    uniform float pointSize;
    uniform float decayModifier;

    varying vec4 vColor;

    float _punctualLightIntensityToIrradianceFactor( const in float lightDistance, const in float cutoffDistance, const in float decayExponent ) {
    	if( decayExponent > 0.0 ) {
    		return pow( saturate( -lightDistance / cutoffDistance + 0.9 ), decayExponent );
    	}
    	return 1.0;
    }

    void main()
    {
      vec4 worldPosition = modelMatrix * vec4(position, 1.0);
      vec4 modelViewPosition = modelViewMatrix * vec4(position, 1.0);
      gl_Position = projectionMatrix * modelViewPosition;

      gl_PointSize = pointSize;
      vec4 addedLights = vec4(0.1, 0.1, 0.1, 1.0);
      for(int l = 0; l < NUM_POINT_LIGHTS; l++) {
        float dist = distance(pointLights[l].position, modelViewPosition.xyz);
        addedLights.rgb += pointLights[l].color *
        _punctualLightIntensityToIrradianceFactor(
            dist,
            pointLights[l].distance * 0.8,
            pointLights[l].decay * decayModifier
          );
      }
      vColor = addedLights;
    }
  `,
  fragmentShader: `
    varying vec4 vColor;

    void main() {
      gl_FragColor = vColor;
    }
  `,
})


src/shaders/VolumetricLightCylinder.js
import * as THREE from 'three'

export default () => ({
  lights: true,
  transparent: true,
  depthWrite: false,
  uniforms: THREE.UniformsUtils.merge([
    THREE.UniformsLib['lights'],
    {
      attenuation: { value: 25.0 },
      anglePower: { value: 10.0 },
      spotPosition: { value: new THREE.Vector3(0, 0, 0) },
    },
  ]),
  vertexShader: `
    #include <common>

    struct PointLight {
      vec3 position;
      vec3 color;
      float distance;
      float decay;
      int shadow;
      float shadowBias;
      float shadowRadius;
      vec2 shadowMapSize;
      float shadowCameraNear;
      float shadowCameraFar;
    };

    uniform PointLight pointLights[NUM_POINT_LIGHTS];
    uniform float viewMix;

    varying vec4 vColor;
    varying vec3 vNormal;
    varying vec3 vWorldPosition;

    float _punctualLightIntensityToIrradianceFactor( const in float lightDistance, const in float cutoffDistance, const in float decayExponent ) {
    	if( decayExponent > 0.0 ) {
    		return pow( saturate( -lightDistance / cutoffDistance + 1.0 ), decayExponent );
    	}
    	return 1.0;
    }

    void main() {
      vNormal = normalize(normalMatrix * normal);

      vec4 worldPosition = modelMatrix * vec4(position, 1.0);
      vec4 modelViewPosition = modelViewMatrix * vec4(position, 1.0);
      gl_Position = projectionMatrix * modelViewPosition;

      vec4 addedLights = vec4(0.1, 0.1, 0.1, 1.0);
      for(int l = 0; l < NUM_POINT_LIGHTS; l++) {
        float dist = distance(pointLights[l].position, worldPosition.xyz);
        addedLights.rgb += pointLights[l].color *
          _punctualLightIntensityToIrradianceFactor(
            dist,
            pointLights[l].distance,
            pointLights[l].decay
          );
      }

      vWorldPosition = worldPosition.xyz;
      vColor = addedLights;
    }
  `,
  fragmentShader: `
    varying vec3 vNormal;
    varying vec3 vWorldPosition;
    uniform vec3 spotPosition;
    uniform float attenuation;
    uniform float anglePower;
    varying vec4 vColor;

    void main() {
      float intensity;
      intensity	= distance(vWorldPosition, spotPosition)/attenuation;
      intensity	= 1.0 - clamp(intensity, 0.0, 1.0);
      vec3 normal	= vec3(vNormal.x, vNormal.y, abs(vNormal.z));
      float angleIntensity	= pow(dot(normal, vec3(0.0, 0.0, 1.0)), anglePower);
      intensity	= intensity * angleIntensity;
      gl_FragColor = vec4(vColor.rgb, intensity);
    }
  `,
})


src/shaders/VolumetricLightScattering.js
import { Vector2 } from 'three'

export default () => ({
  uniforms: {
    tDiffuse: { value: null },
    lightPosition: { value: new Vector2(0.5, 0.5) },
    exposure: { value: 0.2 },
    decay: { value: 0.96 },
    density: { value: 0.6 },
    weight: { value: 0.2 },
    samples: { value: 80 },
  },
  vertexShader: `
    varying vec2 vUv;

    void main() {
      vUv = uv;
      gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
    }
  `,
  fragmentShader: `
    varying vec2 vUv;
    uniform sampler2D tDiffuse;
    uniform vec2 lightPosition;
    uniform float exposure;
    uniform float decay;
    uniform float density;
    uniform float weight;
    uniform int samples;
    const int MAX_SAMPLES = 100;

    void main() {
      vec2 texCoord = vUv;
      vec2 deltaTextCoord = texCoord - lightPosition;
      deltaTextCoord *= 1.0 / float(samples) * density;
      vec4 color = texture2D(tDiffuse, texCoord);
      float a = color.a;
      float illuminationDecay = 1.0;
      for(int i=0; i < MAX_SAMPLES; i++) {
        if(i == samples) {
          break;
        }
        texCoord -= deltaTextCoord;
        vec4 sample = texture2D(tDiffuse, texCoord);
        sample *= illuminationDecay * weight;
        color += sample;
        illuminationDecay *= decay;
      }
      gl_FragColor = color * exposure;
    }
  `,
})


src/index.js
import * as THREE from 'three'
import debounce from 'lodash.debounce'

import { GLTFLoader } from '../three-examples/loaders/GLTFLoader'
import { EffectComposer } from '../three-examples/postprocessing/EffectComposer'
import { RenderPass } from '../three-examples/postprocessing/RenderPass'
import { ShaderPass } from '../three-examples/postprocessing/ShaderPass'
import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls.js'

import AdditiveShader from './shaders/Additive'
import ASCIIShader from './shaders/ASCII'
import RippleShader from './shaders/Ripple'
import ScanShader from './shaders/Scan'
import VertexLitParticle from './shaders/VertexLitParticle'
import VolumetricLightScattering from './shaders/VolumetricLightScattering'
import VolumetricLightCylinder from './shaders/VolumetricLightCylinder'

// Constants

const DEFAULT_LAYER = 0
const OCCLUSION_LAYER = 1

const FONT_MAP_SIZE = new THREE.Vector2(64, 64)
const FONT_CHAR_SIZE = new THREE.Vector2(8, 8)

// Sizes
const sizes = {
  width: window.innerWidth,
  height: window.innerHeight,
}
document.styleSheets[0].insertRule('canvas { outline:none; border:none; }', 0)
// Create Scene + Camera

const mainScene = new THREE.Scene()
const cameraGroup = new THREE.Group()
mainScene.add(cameraGroup)

const mainCamera = new THREE.PerspectiveCamera(
  15,
  window.innerWidth / window.innerHeight,
  8,
  15
  // WIDE ANGLE VIEW
  // 35,
  // sizes.width / sizes.height,
  // 0.1,
  // 100
)
mainCamera.position.z = 12
cameraGroup.add(mainCamera)

const occlusionCamera = mainCamera.clone()
occlusionCamera.layers.set(OCCLUSION_LAYER)

// Add Point Lights

const backLight = new THREE.PointLight(0x00aaff, 8, 10)
backLight.layers.enable(OCCLUSION_LAYER)
backLight.position.set(-5, 5, -5)
mainScene.add(backLight)

const fillLight = new THREE.PointLight(0x00aaff, 5, 10)
fillLight.layers.enable(OCCLUSION_LAYER)
fillLight.position.set(-5, 0, 5)
mainScene.add(fillLight)

const keyLight = new THREE.PointLight(0xff00ff, 3, 10)
keyLight.layers.enable(OCCLUSION_LAYER)
keyLight.position.set(5, 0, 0)
mainScene.add(keyLight)

// Create Renderer

const renderer = new THREE.WebGLRenderer()
// Main Render Size
renderer.setSize(window.innerWidth, window.innerHeight)
document.getElementById('app').appendChild(renderer.domElement)
renderer.setPixelRatio(Math.min(window.devicePixelRatio, 2))

// switch model code
// const modelsArray =[]
// setTimeout(() => {
//   for (let i = 0; i < modelsArray.length; i++) {
//     const modelFile = require( modelsArray[i]);

//   }
// }, 15000);
const loader = new GLTFLoader()
// Load 3D Model
const modelFile = require('../static/model/cybertruck.glb')
const modelContainer = new THREE.Group()
modelContainer.layers.enable(OCCLUSION_LAYER)
mainScene.add(modelContainer)

loader.load(
  modelFile,
  (gltf) => {
    // Add default mesh
    modelContainer.add(gltf.scene)

    // Add black mesh set to occlusion Layer
    const occlusionScene = gltf.scene.clone()
    const blackMaterial = new THREE.MeshBasicMaterial({
      color: new THREE.Color(0x000000),
    })
    occlusionScene.traverse((node) => {
      if (node.material) {
        node.material = blackMaterial
      }
      if (node.layers) {
        node.layers.set(OCCLUSION_LAYER)
      }
    })
    modelContainer.add(occlusionScene)

    // // Redundant
    // gltf.scene.scale.set(0.8, 0.8, 0.8)
  },
  undefined,
  console.error
)

// Other 3D Models
// const material = new THREE.MeshToonMaterial({ color: '#ffeded' })

// const mesh1 = new THREE.Mesh(new THREE.TorusGeometry(1, 0.4, 16, 60), material)
// const mesh2 = new THREE.Mesh(new THREE.ConeGeometry(1, 2, 32), material)
// const mesh3 = new THREE.Mesh(
//   new THREE.TorusKnotGeometry(0.8, 0.35, 100, 16),
//   material
// )
// // mainScene.add(mesh1, mesh2, mesh3)
const objectsDistance = 4
// mesh1.position.y = -objectsDistance * 0
// mesh2.position.y = -objectsDistance * 1
// mesh3.position.y = -objectsDistance * 2

// Generic

function getLowResSize() {
  const charCountPrecise = [
    window.innerWidth / FONT_CHAR_SIZE.x,
    window.innerHeight / FONT_CHAR_SIZE.y,
  ]

  const charCountCeil = charCountPrecise.map(Math.ceil)

  return {
    charCountPrecise,
    charCountCeil,
  }
}

const startingSizeData = getLowResSize()
const lowResRenderTarget = new THREE.WebGLRenderTarget(
  startingSizeData.charCountCeil[0] * 2,
  startingSizeData.charCountCeil[1] * 2
)

const lowResDepthTexture = new THREE.DepthTexture()
lowResDepthTexture.type = THREE.UnsignedShortType
lowResRenderTarget.depthTexture = lowResDepthTexture

const lowResEffectRenderTarget = new THREE.WebGLRenderTarget(
  startingSizeData.charCountCeil[0] * 2,
  startingSizeData.charCountCeil[1] * 2
)

const occlusionRenderTarget = new THREE.WebGLRenderTarget(
  startingSizeData.charCountCeil[0] * 2,
  startingSizeData.charCountCeil[1] * 2
)

// Ripple Effect

const RIPPLE_SPEED = 0.3
const RIPPLE_PEAK = 0.4

const ripples = []
const rippleCanvas = document.createElement('canvas')
rippleCanvas.width = rippleCanvas.style.width = window.innerWidth
rippleCanvas.height = rippleCanvas.style.height = window.innerHeight
const rippleContext = rippleCanvas.getContext('2d')
const rippleTexture = new THREE.Texture(rippleCanvas)
rippleTexture.minFilter = THREE.NearestFilter
rippleTexture.magFilter = THREE.NearestFilter
rippleCanvas
let rippleWasRendering = false

const canvas = document.getElementById('app')

const controls = new OrbitControls(mainCamera, canvas)
// controls.target = modelFile
controls.panSpeed = 0.1
controls.maxDistance = 12.0
controls.minDistance = 12.0
controls.maxPolarAngle = Math.PI / 2
controls.minPolarAngle = Math.PI / 2
controls.rotateSpeed = 0.5

// Ripples
const linear = (t) => t
const easeOutQuart = (t) => 1 - --t * t * t * t

function renderRipples(delta) {
  if (ripples.length) {
    rippleWasRendering = true

    rippleContext.fillStyle = 'rgb(128, 128, 0)'
    rippleContext.fillRect(0, 0, rippleCanvas.width, rippleCanvas.height)

    ripples.forEach((ripple, i) => {
      ripple.age += delta * RIPPLE_SPEED

      if (ripple.age > 1) {
        ripples.splice(i, 1)
        return
      }

      const size = rippleCanvas.height * easeOutQuart(ripple.age)

      const alpha =
        ripple.age < RIPPLE_PEAK
          ? easeOutQuart(ripple.age / RIPPLE_PEAK)
          : 1 - linear((ripple.age - RIPPLE_PEAK) / (1 - RIPPLE_PEAK))

      let grd = rippleContext.createRadialGradient(
        ripple.position.x,
        ripple.position.y,
        size * 0.25,
        ripple.position.x,
        ripple.position.y,
        size
      )

      grd.addColorStop(1, `rgba(128, 128, 0, 0.5)`)
      grd.addColorStop(
        0.8,
        `rgba(${ripple.color.x}, ${ripple.color.y}, ${20 * alpha}, ${alpha})`
      )
      grd.addColorStop(0, `rgba(0, 0, 0, 0)`)

      rippleContext.beginPath()
      rippleContext.fillStyle = grd
      rippleContext.arc(
        ripple.position.x,
        ripple.position.y,
        size,
        0,
        Math.PI * 2
      )
      rippleContext.fill()
    })

    rippleTexture.needsUpdate = true
  } else if (rippleWasRendering) {
    rippleContext.fillStyle = 'rgb(128, 128, 0)'
    rippleContext.fillRect(0, 0, rippleCanvas.width, rippleCanvas.height)

    rippleWasRendering = false
    rippleTexture.needsUpdate = true
  }
}

function addRipple(event) {
  ripples.push({
    age: 0,
    position: new THREE.Vector2(event.clientX, event.clientY),
    color: new THREE.Vector2(
      (event.clientX / window.innerWidth) * 255,
      (event.clientY / window.innerHeight) * 255
    ),
  })
}
window.addEventListener('click', addRipple)

// New Code (Portfolio + Projects)
// let scrollY = window.scrollY
// let currentSection = 0
// window.addEventListener('scroll', () => {
//   scrollY = window.scrollY
//   //   console.log(scrollY)
// })

const particlesCount = 200
const positions = new Float32Array(particlesCount * 3)

for (let i = 0; i < particlesCount; i++) {
  positions[i * 3 + 0] = (Math.random() - 0.5) * 10
  positions[i * 3 + 1] =
    objectsDistance * 0.5 - Math.random() * objectsDistance * 3
  positions[i * 3 + 2] = (Math.random() - 0.5) * 10
}

const particlesGeometry = new THREE.BufferGeometry()
particlesGeometry.setAttribute(
  'position',
  new THREE.BufferAttribute(positions, 3)
)
const particlesMaterial = new THREE.ShaderMaterial(VertexLitParticle())

const particles = new THREE.Points(particlesGeometry, particlesMaterial)
mainScene.add(particles)

// Galaxy Particles Code experiment
// const particleGeometry = new THREE.SphereBufferGeometry(100, 32, 16)

// let _particleSpeeds = []

// for (let i = 0; i < parameters.count; i++) {
//   const x = Math.random() * frustumWidthHalf
//   const y = -Math.random() * frustumHeightHalf
//   const z = (Math.random() * 2 - 1) * (PARTICLE_DEPTH / 2)
//   _particlePositions.push(x, y, z)
//   _particleSpeeds.push(1 + Math.random() * PARTICLE_SPEED)
// }

// const particleSpeeds = new Float32Array(_particleSpeeds)
// const particleStartPositions = new Float32Array(_particlePositions)
// const particlePositions = new THREE.Float32BufferAttribute(
//   _particlePositions,
//   3
// )
// particleGeometry.setAttribute('position', particlePositions)

// const particleMaterial = new THREE.ShaderMaterial(VertexLitParticle())
// particleMaterial.uniforms.pointSize.value = 2.0
// particleMaterial.uniforms.decayModifier.value = 2.5
// const particles = new THREE.Points(particleGeometry, particleMaterial)
// particles.position.set(-5, 0, -8)
// particlesGroup.add(particles)

// const mousePositionNormalized = new THREE.Vector2(0, 0)

// function animateParticles(delta) {
//   let i = 0
//   for (let p = 0; p < parameters.count; p++) {
//     particlesAttribute.array[i] =
//       (particleStartPositions[i] *
//         frustumWidthHalf *
//         (1.0 + mousePositionNormalized.x * delta * 4.0) *
//         0.2) %
//       frustumWidth

//     particlesAttribute.array[i + 1] =
//       (particleStartPositions[i + 1] *
//         frustumHeightHalf *
//         (1.0 - mousePositionNormalized.y * delta * 4.0) *
//         0.1) %
//       frustumHeight

//     i += 3
//   }

//   particlesAttribute.needsUpdate = true
// }

// Volumetric Lighting

const lightGeometry = new THREE.CylinderGeometry(3, 6, 15, 32, 6, true)
lightGeometry.applyMatrix4(
  new THREE.Matrix4().makeTranslation(
    0,
    -lightGeometry.parameters.height / 2,
    0
  )
)
lightGeometry.applyMatrix4(new THREE.Matrix4().makeRotationX(-Math.PI / 2))

const lightCylinderMaterial = new THREE.ShaderMaterial(
  VolumetricLightCylinder()
)
const lightConeTarget = new THREE.Vector3(0, 0, -8)
const lightCone = new THREE.Mesh(lightGeometry, lightCylinderMaterial)
lightCone.position.set(-5, 2, -8)
lightCone.layers.set(OCCLUSION_LAYER)
lightCylinderMaterial.uniforms.spotPosition.value = lightCone.position
mainScene.add(lightCone)

// ASCII Effect

const fontLoader = new THREE.TextureLoader()
const fontFile = require('../font.png')
const tFont = fontLoader.load(fontFile)
tFont.minFilter = THREE.NearestFilter
tFont.magFilter = THREE.NearestFilter

const asciiPass = new ShaderPass(ASCIIShader())
asciiPass.needsSwap = false
asciiPass.uniforms.tLowRes.value = lowResEffectRenderTarget.texture
asciiPass.uniforms.tDepth.value = lowResRenderTarget.depthTexture
asciiPass.uniforms.cameraNear.value = mainCamera.near / 0.6
asciiPass.uniforms.cameraFar.value = mainCamera.far * 0.35
asciiPass.uniforms.tFont.value = tFont

const fontCountX = FONT_MAP_SIZE.x / FONT_CHAR_SIZE.x
const fontCountY = FONT_MAP_SIZE.y / FONT_CHAR_SIZE.y

asciiPass.uniforms.fontCharTotalCount.value =
  Math.floor(fontCountX) * Math.floor(fontCountY)

asciiPass.uniforms.fontCharSize.value.set(1 / fontCountX, 1 / fontCountY)

asciiPass.uniforms.fontCharCount.value.set(fontCountX, fontCountY)

// Occlusion Composer

const occlusionComposer = new EffectComposer(renderer, occlusionRenderTarget)
occlusionComposer.renderToScreen = false

occlusionComposer.addPass(new RenderPass(mainScene, occlusionCamera))

const lightScatteringPass = new ShaderPass(VolumetricLightScattering())
lightScatteringPass.needsSwap = false
occlusionComposer.addPass(lightScatteringPass)

// Effect Composer

const effectComposer = new EffectComposer(renderer, lowResEffectRenderTarget)
effectComposer.renderToScreen = false

const additivePass = new ShaderPass(AdditiveShader())
additivePass.textureID = null
additivePass.uniforms.tDiffuse.value = lowResRenderTarget.texture
additivePass.uniforms.tAdd.value = occlusionRenderTarget.texture
effectComposer.addPass(additivePass)

const scanPass = new ShaderPass(ScanShader())
scanPass.uniforms.tDepth.value = lowResDepthTexture
scanPass.uniforms.cameraNear.value = mainCamera.near
scanPass.uniforms.cameraFar.value = mainCamera.far
effectComposer.addPass(scanPass)

const ripplePass = new ShaderPass(RippleShader())
ripplePass.uniforms.tRipple.value = rippleTexture
ripplePass.needsSwap = false
effectComposer.addPass(ripplePass)

// Final Composer

const finalComposer = new EffectComposer(renderer)
finalComposer.addPass(asciiPass)

// Mouse Move
const cursor = {}
cursor.x = 0
cursor.y = 0

window.addEventListener('mousemove', (event) => {
  cursor.x = event.clientX / sizes.width - 0.5
  cursor.y = event.clientY / sizes.height - 0.5
  lightCone.position.x = 5 * ((event.clientX / window.innerWidth) * 2 - 1)
  backLight.position.x = lightCone.position.x
})
function updateAsciiRenderSize() {
  const size = getLowResSize()

  asciiPass.uniforms.renderCharSize.value.set(
    1 / size.charCountPrecise[0],
    1 / size.charCountPrecise[1]
  )

  asciiPass.uniforms.renderCharCount.value.set(
    size.charCountPrecise[0],
    size.charCountPrecise[1]
  )

  lowResRenderTarget.setSize(
    size.charCountCeil[0] * 2,
    size.charCountCeil[1] * 2
  )

  effectComposer.setSize(size.charCountCeil[0] * 2, size.charCountCeil[1] * 2)

  occlusionComposer.setSize(
    size.charCountCeil[0] * 2,
    size.charCountCeil[1] * 2
  )
}

// Handle Window Resize
function resizeRenderer() {
  if (window.innerWidth <= 550) {
    rippleCanvas.width = rippleCanvas.style.width = window.innerWidth / 2
    rippleCanvas.height = rippleCanvas.style.height = window.innerHeight / 2
    canvas.width = canvas.style.width = window.innerWidth / 2
    canvas.height = canvas.style.height = window.innerHeight / 2
    updateAsciiRenderSize()
    // Canvas Height Control
    renderer.setSize(window.innerWidth, window.innerHeight)
    mainCamera.aspect = window.innerWidth / window.innerHeight
    mainCamera.updateProjectionMatrix()
    occlusionCamera.aspect = mainCamera.aspect
    occlusionCamera.updateProjectionMatrix()
    modelContainer.scale.set(0.57, 0.57, 0.57)
    modelContainer.position.set(0, 0, 0.4)
    console.log(modelContainer.position)
  } else if (window.innerWidth > 500 && window.innerWidth <= 854) {
    rippleCanvas.width = rippleCanvas.style.width = window.innerWidth / 2
    rippleCanvas.height = rippleCanvas.style.height = window.innerHeight / 2
    canvas.width = canvas.style.width = window.innerWidth / 2
    canvas.height = canvas.style.height = window.innerHeight / 2
    updateAsciiRenderSize()
    // Canvas Height Control
    renderer.setSize(window.innerWidth, window.innerHeight)
    mainCamera.aspect = window.innerWidth / window.innerHeight
    mainCamera.updateProjectionMatrix()
    occlusionCamera.aspect = mainCamera.aspect
    occlusionCamera.updateProjectionMatrix()
    modelContainer.scale.set(0.6, 0.6, 0.6)
    modelContainer.position.set(0, 0, 0.4)
  } else if (window.innerWidth > 854) {
    rippleCanvas.width = rippleCanvas.style.width = window.innerWidth
    rippleCanvas.height = rippleCanvas.style.height = window.innerHeight
    canvas.width = canvas.style.width = window.innerWidth
    canvas.height = canvas.style.height = window.innerHeight
    updateAsciiRenderSize()
    // Canvas Height Control
    renderer.setSize(window.innerWidth, window.innerHeight)
    mainCamera.aspect = window.innerWidth / window.innerHeight
    mainCamera.updateProjectionMatrix()
    occlusionCamera.aspect = mainCamera.aspect
    occlusionCamera.updateProjectionMatrix()
    modelContainer.scale.set(0.8, 0.8, 0.8)
    modelContainer.position.set(0, 0.3, 0.4)
  }
}
window.addEventListener('resize', debounce(resizeRenderer, 50))

// Render Scene

const clock = new THREE.Clock()
let previousTime = 0

modelContainer.rotation.x = 0.1
mainCamera.lookAt(modelContainer.position)

console.log(modelContainer.position)

resizeRenderer()
function render() {
  const elapsedTime = clock.getElapsedTime()
  const delta = elapsedTime - previousTime
  previousTime = elapsedTime

  // Animate Camera (When Projects Sections Added)
  // mainCamera.position.y = (-scrollY / sizes.height) * objectsDistance

  const parallaxX = cursor.x * 1.5
  const parallaxY = -cursor.y * 0.5
  cameraGroup.position.x += (parallaxX - cameraGroup.position.x) * 5 * delta
  cameraGroup.position.y += (parallaxY - cameraGroup.position.y) * 5 * delta

  // Object Animation
  modelContainer.rotation.y += delta / 2.2
  // Drifting pixels - Speed Now Reduced.(Epilepsy Measure)
  particles.rotation.y += delta * 0.3
  // particlesGroup.position.x += -delta / 0.5

  // Scan
  scanPass.uniforms.scan.value =
    (scanPass.uniforms.scan.value + delta * 0.5) % 2

  // Volumetric Lighting

  lightCone.lookAt(lightConeTarget)
  lightCylinderMaterial.uniforms.spotPosition.value = lightCone.position
  const lightConePosition = lightCone.position.clone()
  const vector = lightConePosition.project(occlusionCamera)
  lightScatteringPass.uniforms.lightPosition.value.set(
    (vector.x + 1) / 2,
    (vector.y + 1) / 2
  )

  // Render

  renderRipples(delta)

  renderer.setRenderTarget(lowResRenderTarget)
  renderer.render(mainScene, mainCamera)

  renderer.setRenderTarget(occlusionRenderTarget)
  occlusionComposer.render()

  renderer.setRenderTarget(lowResEffectRenderTarget)
  effectComposer.render()

  renderer.setRenderTarget(null)
  finalComposer.render()
  // animateParticles(delta)
  requestAnimationFrame(render)
}
render()
